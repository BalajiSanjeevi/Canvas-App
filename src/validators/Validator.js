import {VALID_COMMANDS} from '../constants/CommonConstants';
import {OUT_OF_BOUNDARY_ERR_MSG, NOT_ENOUGH_PARAMS_ERR_MSG,
    NEGATIVE_COORDINATES_ERR_MSG, NAN_ERR_MSG, INVALID_COMMAND_ERR_MSG} from '../constants/ErrorConstants';

/*
 * validates the user command against the available list of commands whether
 * it exist in the list and throws error if doesnt exist.
 */
export function validateCommand (command) {
    if(!VALID_COMMANDS.includes(command))
        throw Error(INVALID_COMMAND_ERR_MSG);
    return true;
}

/*
 * validates the given command sets against the given expected length and throws error if the length doesn't match.
 */
export function validateCommandSetLength(commandSet, expectedLength) {
    if(commandSet.length !== expectedLength)
        throw Error(NOT_ENOUGH_PARAMS_ERR_MSG);
    return true;
}

/*
 * Validates the given input whether its number or not. Throws error when any other character is passed other than number.
 */
export function isNumber(value) {
    if(isNaN(value))
        throw Error(NAN_ERR_MSG);
    return true;
}

/*
 * Validates the given input for number and should be greater than 0.
 */
export function validateNumber(value) {
    if(isNumber(value) && parseInt(value) <= 0)
        throw Error(NEGATIVE_COORDINATES_ERR_MSG);
    return true;
}

/*
 * Validates given array of input for number.
 */
export function validateNumbers(values) {
    values.forEach((char) => {
        validateNumber(char);
    });
    return true;
}

/*
 * Validates the given input for its lenght against the expected length and checks whether all characters are number.
 */
export function validateCommandSets(commandSets, expectedCommandLength){
    validateCommandSetLength(commandSets, expectedCommandLength);
    return validateNumbers(commandSets);
}

/*
 * Validates the given pixel whether it falls within the canvas boundary.
 */
export function isCoordinatesWithinCanvasBoundary(canvas, pixels) {
    pixels.forEach((pixel) => {
        if(pixel.axisY < 1 || pixel.axisY > canvas.pixels.length -2
            || pixel.axisX < 1|| pixel.axisX > canvas.pixels[0].length -2)
            throw Error(OUT_OF_BOUNDARY_ERR_MSG);
    });
    return true;
}
