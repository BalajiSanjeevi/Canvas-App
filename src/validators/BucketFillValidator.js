import * as validate from './Validator';
import {FILL_COLOUR_CHARACTER_INVALID_LENGTH_ERR_MSG} from '../constants/ErrorConstants';

const EXPECTED_INPUT_LENGTH = 2;

/*
 * The fill character should be 1 character and should be defined.
 */
function validateFillChar(fillChar){
    if(!fillChar || fillChar.length !== 1)
        throw Error(FILL_COLOUR_CHARACTER_INVALID_LENGTH_ERR_MSG);
}
/*
 * Checks whether the coordinate pair falls within the boundary.
 */
function validateCoordinates([axisX, axisY], canvas) {
    return validate.isCoordinatesWithinCanvasBoundary(canvas, [{axisX: axisX, axisY: axisY}]);
}

/*
 * validates the given command set for the set of validations to check whether it satisfy for bucket filling.
 * 1. Given input should have 2 characters
 * 2. Both characters should be a number
 * 3. The coordinate pair should fall with in the boundary.
 * 4. The fill character should be single character.
 */
export default ([axisX, axisY, fillColor], canvas) => {
    validate.validateCommandSets([axisX, axisY], EXPECTED_INPUT_LENGTH);
    validateFillChar(fillColor);
    const params = [axisX, axisY].map((value) => parseInt(value));
    validateCoordinates(params, canvas);
    return true;
}