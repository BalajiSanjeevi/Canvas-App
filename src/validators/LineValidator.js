import * as validate from './Validator';
import {INVALID_COORDINATES_PAIR_LINE_ERR_MSG} from '../constants/ErrorConstants';

const EXPECTED_INPUT_LENGTH = 4;


/*
 * Checks whether the given coordinate pair satisfy for drawing either vertical or horizontal line.
 */
function isValidCoordinates([axisX1, axisY1, axisX2, axisY2]){
    if(!(axisX1 === axisX2 || axisY1 === axisY2))
        throw Error(INVALID_COORDINATES_PAIR_LINE_ERR_MSG);
    return true;
}

/*
 * Checks whether the given coordinate pair satisfy for drawing either vertical or horizontal line and falls within the canvas boundary.
 */

function validateCoordinates(commandSets, canvas){
    let pixels = [{axisX: commandSets[0], axisY: commandSets[1]}, {axisX: commandSets[2], axisY: commandSets[3]}];
    return isValidCoordinates(commandSets) && validate.isCoordinatesWithinCanvasBoundary(canvas, pixels);
}

/*
 * validates the given command set for the set of validations to check whether it satisfy for creating the Line.
 * 1. Given input should have 4 characters
 * 2. All 4 characters should be a number
 * 3. The coordinate pair should satisfy for drawing either vertical or horizontal line
 * 4. The coordinate pairs should fall with in the boundary.
 */

export default (commandSets, canvas) => {
    validate.validateCommandSets(commandSets, EXPECTED_INPUT_LENGTH);
    const params = commandSets.map((value) => parseInt(value));
    validateCoordinates(params, canvas);
    return true;
}