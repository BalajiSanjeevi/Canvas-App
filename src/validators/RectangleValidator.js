import * as validate from './Validator';
import {INVALID_COORDINATES_PAIR_RECTANGLE_ERR_MSG} from '../constants/ErrorConstants';

const EXPECTED_INPUT_LENGTH = 4;

/*
 * checks whether the given coordinate pair satisfy for creating the rectangle.
 */
function isValidCoordinates([axisX1, axisY1, axisX2, axisY2]){
    if(axisX2 <= axisX1 || axisY2 <= axisY1)
        throw Error(INVALID_COORDINATES_PAIR_RECTANGLE_ERR_MSG);
    return true;
}
/*
 * checks whether given coordinate pairs satisfy for creating rectangle and falls within the canvas boundary.
 */
function validateCoordinates(commandSets, canvas){
    let pixels = [{axisX: commandSets[0], axisY: commandSets[1]}, {axisX: commandSets[2], axisY: commandSets[3]}];
    return isValidCoordinates(commandSets) && validate.isCoordinatesWithinCanvasBoundary(canvas, pixels);
}

/*
 * validates the given command set for the set of validations to check whether it satisfy for creating the rectangle.
 * 1. Given input should have 4 characters
 * 2. All 4 characters should be a number
 * 3. The coordinate pairs should fall with in the boundary.
 */
export default (commandSets, canvas) => {
    validate.validateCommandSets(commandSets, EXPECTED_INPUT_LENGTH);
    const params = commandSets.map((value) => parseInt(value));
    validateCoordinates(params, canvas);
    return true;
}