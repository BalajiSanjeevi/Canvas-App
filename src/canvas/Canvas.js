import write from '../io/OutputWriter';
import {validateNumbers, validateCommandSetLength} from '../validators/Validator';

const EXPECTED_INPUT_LENGTH = 2;

const BOUNDARY_BORDER_PIXEL_LENGTH = 2;

const LEFT_AND_RIGHT_CANVAS_BORDER_CHAR = "|";

const TOP_AND_BOTTOM_CANVAS_BORDER_CHAR = "-";


/*
 * Canvas class is meant to hold the canvas context which will have the canvas boundary within which we can play around
 * and it has couple of function to help in drawing line and getting the context.
 */
export default class Canvas {

    /*
     * Validates the commandset for the required length and checks whether the command set is all numbers.
     */
    static validate = function (commandSet){
        return  validateCommandSetLength(commandSet, EXPECTED_INPUT_LENGTH) && validateNumbers(commandSet);
    };

    /*
     * Constructor to create the canvas object and initializes the canvas boundary.
     */
    constructor([width, height]){
        this._pixels = [[]];
        this._width = parseInt(width);
        this._height =  parseInt(height);
        this.initializeCanvasBoundary(this._height + BOUNDARY_BORDER_PIXEL_LENGTH, this._width + BOUNDARY_BORDER_PIXEL_LENGTH);
    }

    /*
     * Helps in initializing the canvas boundary.
     */
    initializeCanvasBoundary(height, width){
        this._pixels = [... new Array(height)].map((row, i) => {
            return [... new Array(width)].map((column, j) => {
                if(i === 0 || i === height - 1){
                    return TOP_AND_BOTTOM_CANVAS_BORDER_CHAR;
                } else if(j === 0 || j === width - 1) {
                    return LEFT_AND_RIGHT_CANVAS_BORDER_CHAR;
                }
                return " ";
            });
        });
    }

    /*
     * Exposes the pixels which contains the actual canvas boundary with the state of it.
     */
    get pixels(){
        return this._pixels;
    }

    /*
     *  Calls the write function to print all the rows and columns.
     */
    draw(){
        this._pixels.forEach((row) => {
            write(row.join(""));
        })
    }

    /*
     * Fills the character 'X' from the given startPoint and endPoint index in the given axisY row
     */
    drawHorizontalLine(startPoint, endPoint, axisY){
        // As per Fill function, it fills the character until < endPoint. As we need to include the last position like (<=endPoint). so adding 1 to the endpoint.
        this._pixels[axisY].fill('X', startPoint, endPoint + 1);
    }
    /*
     * Fills the character 'X' on each row from the given startPoint and endPoint index at the given axisX column index.
     */
    drawVerticalLine(startPoint, endPoint, axisX){
        for(var i = startPoint; i <= endPoint; i++){
            this._pixels[i][axisX] = 'X';
        }
    }
}