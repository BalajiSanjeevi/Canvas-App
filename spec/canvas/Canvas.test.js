import Canvas from '../../src/canvas/Canvas';
import * as validate from '../../src/validators/Validator';
import writer from '../../src/io/OutputWriter';

jest.mock('../../src/validators/Validator');
jest.mock('../../src/io/OutputWriter');

describe("Canvas Test", () => {

    it("should initialize the canvas boundary", () => {
        const canvas = new Canvas([10, 10]);
        expect(canvas.pixels).toMatchSnapshot();
    });

    describe("validate", () => {

        it("should return true when input [1, 2]", () => {
            validate.validateCommandSetLength.mockReturnValue(true);
            validate.validateNumbers.mockReturnValue(true);
            expect(Canvas.validate(['1', '2'])).toBeTruthy();
        });

        it("should return false when input [1, 2, 3]", () => {
            validate.validateCommandSetLength.mockReturnValue(false);
            validate.validateNumbers.mockReturnValue(true);
            expect(Canvas.validate(['1', '2', '3'])).toBeFalsy();
        });

        it("should have called validateCommandSetLength with [3, 4] and 2", () => {
            Canvas.validate(['3', '4']);
            expect(validate.validateCommandSetLength).toBeCalledWith(['3', '4'], 2);
        });

        it("should have called validateNumber with [1,2]", () => {
            Canvas.validate(['1', '2']);
            expect(validate.validateNumbers).toBeCalledWith(['1', '2']);
        });
    });

    describe("Canvas core functions test", () => {
        let canvas;
        beforeEach(() => {
            canvas = new Canvas([10, 10]);
        });

        it("should fill row 2 with character x from index 1 to 6 when drawHorizontalLine called", ()=> {
            canvas.drawHorizontalLine(1, 6, 2);
            const row2 = canvas.pixels[2];
            row2.forEach((char, i) => {
                if(i >= 1 && i <= 6)
                    expect(row2[i]).toBe('X');
            });
        });

        it("should fill column 2 with character x from row 1 to 6 when drawVerticalLine called", ()=> {
            canvas.drawVerticalLine(1, 6, 2);

            canvas.pixels.forEach((row, i) => {
                if(i >= 1 && i <= 6)
                    expect(row[2]).toBe('X');
            });
        });

        it("draw function should call the write function 12 times with the each row as string", ()=> {
            canvas.draw();
            expect(writer.mock.calls.length).toBe(12);
            canvas.pixels.forEach((row, i) => {
                expect(writer.mock.calls[i].toString()).toBe(row.join(""));
            });
        })
    });

});