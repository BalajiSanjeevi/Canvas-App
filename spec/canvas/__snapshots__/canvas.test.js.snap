// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas Test should initialize the canvas boundary 1`] = `
------------
|          |
|          |
|          |
|          |
|          |
|          |
|          |
|          |
|          |
|          |
------------
`;
